chia/util/streamable.py

```diff
@@ -298,7 +246,158 @@ def stream_str(item: Any, f: BinaryIO) -> None:
+
+    stream_functions = []
+    parse_functions = []
+    try:
+        hints = get_type_hints(cls)
+        fields = {field.name: hints.get(field.name, field.type) for field in dataclasses.fields(cls)}
+    except Exception:
+        fields = {}
+
+    FIELDS_FOR_STREAMABLE_CLASS[cls] = fields
+
+    for _, f_type in fields.items():
+        stream_functions.append(cls.function_to_stream_one_item(f_type))
+        parse_functions.append(cls.function_to_parse_one_item(f_type))
+
+    STREAM_FUNCTIONS_FOR_STREAMABLE_CLASS[cls] = stream_functions
+    PARSE_FUNCTIONS_FOR_STREAMABLE_CLASS[cls] = parse_functions
+    return cls
+
+
 class Streamable:
+    """
+    This class defines a simple serialization format, and adds methods to parse from/to bytes and json. It also
+    validates and parses all fields at construction in ´__post_init__` to make sure all fields have the correct type
+    and can be streamed/parsed properly.
+
+    The available primitives are:
+    * Sized ints serialized in big endian format, e.g. uint64
+    * Sized bytes serialized in big endian format, e.g. bytes32
+    * BLS public keys serialized in bls format (48 bytes)
+    * BLS signatures serialized in bls format (96 bytes)
+    * bool serialized into 1 byte (0x01 or 0x00)
+    * bytes serialized as a 4 byte size prefix and then the bytes.
+    * ConditionOpcode is serialized as a 1 byte value.
+    * str serialized as a 4 byte size prefix and then the utf-8 representation in bytes.
+
+    An item is one of:
+    * primitive
+    * Tuple[item1, .. itemx]
+    * List[item1, .. itemx]
+    * Optional[item]
+    * Custom item
+
+    A streamable must be a Tuple at the root level (although a dataclass is used here instead).
+    Iters are serialized in the following way:
+
+    1. A tuple of x items is serialized by appending the serialization of each item.
+    2. A List is serialized into a 4 byte size prefix (number of items) and the serialization of each item.
+    3. An Optional is serialized into a 1 byte prefix of 0x00 or 0x01, and if it's one, it's followed by the
+       serialization of the item.
+    4. A Custom item is serialized by calling the .parse method, passing in the stream of bytes into it. An example is
+       a CLVM program.
+
+    All of the constituents must have parse/from_bytes, and stream/__bytes__ and therefore
+    be of fixed size. For example, int cannot be a constituent since it is not a fixed size,
+    whereas uint32 can be.
+
+    Furthermore, a get_hash() member is added, which performs a serialization and a sha256.
+
+    This class is used for deterministic serialization and hashing, for consensus critical
+    objects such as the block header.
+
+    Make sure to use the streamable decorator when inheriting from the Streamable class to prepare the streaming caches.
+    """
+
+    def post_init_parse(self, item: Any, f_name: str, f_type: Type) -> Any:
```

Comment:
This is just a function, right?  It only uses `self` to access itself for recursive calls.

Reply:
Yeah well its just what we had in `type_checking.py` (ok, with a different name haha) but i have plans for it in a future PR anyway so i would just leave it as is here.